README.md
# Forward glance statistics as gelf logs

This command line utility forwards [Glances] statistics and forwards the data
to [Graylog].  The program runs in a loop for an (optional) given number of
times in a timeout.  Each loop cycle collects data from the [Glances] server
and forwards the data as a "first class" (descrete fields) Graylog message.

With this data you can configured dashboards to report statistics based on
system health, temperatures etc.

The program currently records the following:

* CPU temperature (sensors)
* CPU load

Other data from [Glances] is very easy to add, I just haven't had a lot of time
to add them but I'd be glad to if someone asked for more.  If so, create an
issue and I'll look into your request.

![Screenshot](doc/snap.png?raw=true "Dashboard Example")

## Installing

You need Python 3 to run the program and the *pip* installer.

Install the library and program:
```bash
pip install gelfglance
```

For *systemd* operating systems, the example [systemd service] file might be
useful.


## Usage

Example: forward [Glances] statistics on server `jove.com` at port 61207 to
the [Graylog] server `ossage.com` every minute.

```bash
gelfglance server --gelfserver jove.com --glancesport 61207 --source ossage --glancesserver ossage.com -t 60
```

Here's the usage for convenience (use `--help`):

```pony
Usage: gelfglance <list|...> [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -w NUMBER, --whine=NUMBER
                        add verbosity to logging
  -x, --short           short output for list


Usage: gelfglance server [additional options]

Start in server mode

Options:
  -g STRING, --gelfserver=STRING
                        Graylog host name (defaults to localhost)
  -p INTEGER, --gelfport=INTEGER
                        Graylog port (defaults to 12201)
  -s STRING, --glancesserver=STRING
                        Glances server port (defaults to localhost)
  -l INTEGER, --glancesport=INTEGER
                        Glances port (defaults to 61208)
  -r STRING, --source=STRING
                        Graylog port (defaults to 12201)
  -c STRING, --count=STRING
                        Number of messages to send (default to infinite)
  -t STRING, --timeout=STRING
                        Timeout between each sent message in seconds (defaults to 10)
  -h, --help            show this help message and exit
```


## Changelog

An extensive changelog is available [here](CHANGELOG.md).


## License

Copyright © 2018 Paul Landes

Apache License version 2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


<!-- links -->

[Glances]: https://nicolargo.github.io/glances/
[Graylog]: https://www.graylog.org
[systemd service]: config/glancesgelf.service


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased]

## [0.0.1] - 2018-05-21
### Added
- Initial version

[Unreleased]: https://github.com/plandes/gelfglance/compare/v0.0.1...HEAD
[0.0.2]: https://github.com/plandes/gelfglance/compare/v0.0.1...v0.0.2


src/python/zensols/gelfglance/log.py
import logging
import syslog
import time
import traceback
from gelfclient import UdpClient
from zensols.gelfglance import GlancesClient

logger = logging.getLogger('zensols.gelfglance.log')

class LogSender(object):
    """
    Get glances data from server and forward it on to Graylog 2.

    :gelf_server: the Graylog host name
    :gelf_port: the Graylog port (defaults to 12201)
    :source: used in the source field of the gelf message (defaults to host name)
    :timeout_sec: timeout in seconds before the Glances data is queried and sent to graylog (defaults to 10)
    :count: number of messages to send (defaults to infinity)
    """
    def __init__(self, gelf_server='localhost', gelf_port=12201, source=None,
                 glances_server='localhost', glances_port=61208,
                 timeout_secs=10, count=10):
        self.gelf_server = gelf_server
        self.gelf_port = gelf_port
        self.glances_server = glances_server
        self.glances_port = glances_port
        self.source = source
        self.count = count
        self.timeout_secs = timeout_secs
        self.mtu = 8000

    @property
    def gelf(self):
        "Return the Graylog logger client."
        if not hasattr(self, '_gelf'):
            self._gelf = UdpClient(self.gelf_server, port=self.gelf_port, mtu=self.mtu, source=self.source)
        return self._gelf

    @property
    def glances(self):
        "Return the Glances client."
        if not hasattr(self, '_glances'):
            self._glances = GlancesClient(self.glances_server, self.glances_port)
        return self._glances

    def send_message(self, message, additional_data, level=syslog.LOG_INFO):
        "Send a Gelf message with Glances statistics."
        data = {'short_message': message,
                'program': 'glances',
                'level': level}
        if self.source:
            data['host'] = self.source
        data.update(additional_data)
        logger.debug('sending message to {}:{} -> <{}>'.
                     format(self.gelf_server, self.gelf_port, additional_data))
        self.gelf.log(data)

    def collect_and_send(self):
        "Collect glances data and send to Graylog."
        data = self.glances.get()
        msg = 'System statistics for {}'.format(self.glances.format_system())
        self.send_message(msg, data)

    def start(self):
        "Start the logger as a /server/ by looping on a timeout."
        logger.info('starting gelf_server {gelf_server} on {gelf_port} from source \'{source}\', timeout(s)={timeout_secs}, count={count}'.format(**vars(self)))
        while self.count is None or self.count > 0:
            logger.debug('timeout_secs ({}) up'.format(self.timeout_secs))
            try:
                self.collect_and_send()
            except Exception as e:
                traceback.print_exc()
            if not self.count is None:
                self.count -= 1
            time.sleep(self.timeout_secs)


src/python/zensols/gelfglance/glances.py
import logging
import math
import json
from xmlrpc import client

logger = logging.getLogger('zensols.gelfglance.glances')

class GlancesClient(object):
    """
    Make the xmlrpc request to the glances server and facade out the response.

    See https://github.com/nicolargo/glances/wiki/The-Glances-2.x-API-How-to
    """
    def __init__(self, host='localhost', port=61208, key_prefix='glances'):
        self.host = host
        self.port = port
        self.key_prefix = key_prefix

    @property
    def proxy(self):
        "Return the glances xmlrpc proxy object."
        if not hasattr(self, '_proxy'):
            self._proxy = client.ServerProxy('http://{}:{}'.format(self.host, self.port))
        return self._proxy

    @property
    def plugins(self):
        "Rerturn glances plugins."
        res = self.proxy.getAllPlugins()
        return json.loads(res)

    @property
    def sensors(self):
        "Return sensor data."
        res = self.proxy.getSensors()
        return json.loads(res)

    @property
    def cpu(self):
        "Return CPU data."
        res = self.proxy.getCpu()
        return json.loads(res)

    @property
    def system(self):
        "Return system information (host name, OS data)."
        res = self.proxy.getSystem()
        return json.loads(res)

    def format_system(self):
        "Return a human readable system summary for the message."
        return '{hostname} ({hr_name})'.format(**self.system)

    def _col_sensors(self, sensors, data):
        "Collect sensor info in `data`."
        for td in sensors:
            prefix = '{}_sensor_'.format(self.key_prefix)
            name = prefix + td['label'].replace(' ', '_').lower()
            val = td['value']
            # argument '--fahrenheit' has no affect
            if td['unit'] == 'C':
                val = 9.0/5.0 * val + 32
                # round errors
                val = math.floor(val * 100) / 100
            data[name] = val

    def _col(self, gdict, name, data):
        "Return general data in `gdict` for category `name` in dict `data`."
        for k, v in gdict.items():
            k = '{}_{}_{}'.format(self.key_prefix, name, k)
            data[k] = v

    def get(self, plugins=None):
        "Return data for specifid plugins (defaults to sensors, cpu)."
        if plugins is None:
            plugins = {'sensors', 'cpu'}
        data = {}
        if 'sensors' in plugins:
            self._col_sensors(self.sensors, data)
        if 'cpu' in plugins:
            self._col(self.cpu, 'cpu', data)
        return data


src/python/zensols/gelfglance/cli.py
from zensols.actioncli import OneConfPerActionOptionsCli
from zensols.actioncli import SimpleActionCli
from zensols.gelfglance import LogSender

VERSION='0.1'

# recommended app command line
class ConfAppCommandLine(OneConfPerActionOptionsCli):
    def __init__(self):
        server_action = 'server'
        cnf = {'executors':
               [{'name': 'hello',
                 'executor': lambda params: LogSender(**params),
                 'actions':[{'name': server_action,
                             'meth': 'start',
                             'doc': 'start in server mode',
                             'opts': [['-g', '--gelfserver', True,
                                       {'dest': 'gelf_server', 'metavar': 'STRING',
                                        'default': 'localhost',
                                        'help': 'Graylog host name (defaults to localhost)'}],
                                      ['-p', '--gelfport', True,
                                       {'dest': 'gelf_port', 'metavar': 'INTEGER',
                                        'default': 12201,
                                        'type': 'int',
                                        'help': 'Graylog port (defaults to 12201)'}],
                                      ['-s', '--glancesserver', True,
                                       {'dest': 'glances_server', 'metavar': 'STRING',
                                        'default': 'localhost',
                                        'help': 'Glances server port (defaults to localhost)'}],
                                      ['-l', '--glancesport', True,
                                       {'dest': 'glances_port', 'metavar': 'INTEGER',
                                        'default': 61208,
                                        'type': 'int',
                                        'help': 'Graylog port (defaults to 61208)'}],
                                      ['-r', '--source', True,
                                       {'metavar': 'STRING',
                                        'default': 'glances',
                                        'help': 'Glances port (defaults to 12201)'}],
                                      ['-c', '--count', False,
                                       {'metavar': 'STRING',
                                        'type': 'int',
                                        'help': 'Number of messages to send (default to infinite)'}],
                                      ['-t', '--timeout', True,
                                       {'metavar': 'STRING',
                                        'dest': 'timeout_secs',
                                        'default': 10,
                                        'type': 'int',
                                        'help': 'Timeout between each sent message in seconds (defaults to 10)'}]]
                 }]}],
               'whine': 1}
        super(ConfAppCommandLine, self).__init__(cnf, version=VERSION, default_action=server_action)

    def config_parser(self):
        super(ConfAppCommandLine, self).config_parser()
        parser = self.parser
        parser.remove_option('-s')
        parser.add_option('-x', '--short', dest='short',
                          help='short output for list', action='store_true')

def main():
    cl = ConfAppCommandLine()
    cl.invoke()
